toolkit/primitives/block-rewards-payouts/src/lib.rs

```
@@ -0,0 +1,215 @@
+				}
+			}
+			block_production_payout_info.production_rundowns.sort();
+
+			for rundown in &mut block_production_payout_info.production_rundowns {
+				rundown.beneficiaries.sort()
+			}
+
+			println!("payout time!  {block_production_payout_info:?}");
+
+			Ok(Self { block_production_data: Some(block_production_payout_info) })
+		}
+	}
+
+	fn count_by_epoch_and_producer<BlockProducerId: Eq + Hash>(
+		slot_producers: Vec<(Slot, BlockProducerId)>,
+		mc_epoch_config: &MainchainEpochConfig,
+		slot_duration: SlotDuration,
+	) -> HashMap<McEpochNumber, HashMap<BlockProducerId, u32>> {
+		let mut epoch_producers: HashMap<McEpochNumber, HashMap<BlockProducerId, u32>> =
+			HashMap::new();
+
+		for (slot, producer) in slot_producers {
+			let mc_epoch = mc_epoch_config
+				.timestamp_to_mainchain_epoch(Timestamp::from_unix_millis(
+					slot.timestamp(slot_duration)
+						.expect("Timestamp for past slots can not overflow")
+						.as_millis(),
+				))
+				.expect("Mainchain epoch for past slots exists");
+			let mc_epoch = McEpochNumber(mc_epoch.0 - 2);
+
+			*epoch_producers.entry(mc_epoch).or_default().entry(producer).or_default() += 1;
+		}
+
+		epoch_producers
+	}
+
+	#[async_trait::async_trait]
+	impl<BlockProducerId, BeneficiaryId> InherentDataProvider
+		for BlockRewardsPayoutsInherentDataProvider<BlockProducerId, BeneficiaryId>
+	where
+		BeneficiaryId: Encode + Send + Sync,
+		BlockProducerId: Encode + Send + Sync,
+	{
+		async fn provide_inherent_data(
+			&self,
+			inherent_data: &mut InherentData,
+		) -> Result<(), sp_inherents::Error> {
+			if let Some(block_production_data) = &self.block_production_data {
+				inherent_data.put_data(INHERENT_IDENTIFIER, block_production_data)?
+			}
+			Ok(())
+		}
+
+		async fn try_handle_error(
+			&self,
+			identifier: &InherentIdentifier,
+			mut error: &[u8],
+		) -> Option<Result<(), sp_inherents::Error>> {
+			if *identifier == INHERENT_IDENTIFIER {
+				let error = InherentError::decode(&mut error).ok()?;
```

If the decoding fails this fails silently. Maybe you want to match this on `Ok` and log an error if there was a problem?

---

Added logging